src/sinks/datadog/logs.rs

```diff
@@ -256,40 +272,58 @@ impl HttpSink for DatadogLogsJsonService {
                 count: events.len(),
             });
         }
-        self.config
-            .build_request(self.uri.as_str(), "application/json", body)
+        self.config.build_request(
+            self.uri.as_str(),
+            api_key.as_str(),
+            "application/json",
+            body,
+        )
     }
 }
 
 #[async_trait::async_trait]
 impl HttpSink for DatadogLogsTextService {
-    type Input = Bytes;
-    type Output = Vec<Bytes>;
+    // type Input = Bytes;
+    // type Output = Vec<Bytes>;
+
+    type Input = PartitionInnerBuffer<Bytes, String>;
+    type Output = PartitionInnerBuffer<Vec<Bytes>, String>;
 
     fn encode_event(&self, event: Event) -> Option<EncodedEvent<Self::Input>> {
+        let api_key = event
+            .metadata()
+            .datadog_api_key()
+            .to_owned()
+            .unwrap_or_else(|| self.config.api_key.clone());
+
         encode_event(event, &self.config.encoding).map(|e| {
             emit!(DatadogLogEventProcessed {
                 byte_size: e.item.len(),
                 count: 1,
             });
-            EncodedEvent::new(e.item)
+            EncodedEvent::new(PartitionInnerBuffer::new(e.item, api_key))
         })
     }
 
-    async fn build_request(&self, events: Self::Output) -> crate::Result<http::Request<Vec<u8>>> {
+    async fn build_request(&self, events: Self::Output) -> crate::Result<Request<Vec<u8>>> {
+        let (events, api_key) = events.into_parts();
         let body: Vec<u8> = events.into_iter().flat_map(Bytes::into_iter).collect();
+
         self.config
-            .build_request(self.uri.as_str(), "text/plain", body)
+            .build_request(self.uri.as_str(), api_key.as_str(), "text/plain", body)
```

Comment:
Given that we pass in a reference to the key here, is there a way we can reduce all the cloning going on up to this point? It seems like a fairly excessive amount of cloning is added while passing around/merging API keys.